feat: Add new `includeIgnoreFile()` to config-helpers (#430)
eslint/rewrite: packages/config-helpers/README.md

Reviewer: As I recall, there was a mention that it might be a good opportunity to document `convertIgnorePatternToMinimatch()`; although it isn’t exported from `eslint/config`, it’s available in `@eslint/config-helpers`.

https://github.com/eslint/rewrite/issues/329#issuecomment-4186746147

If possible, could we also add documentation for `convertIgnorePatternToMinimatch()`?

Reviewer 2: It's intentionally undocumented (https://github.com/eslint/rewrite/pull/430#discussion_r3058800836).

]);
```

### `includeIgnoreFile()`

The `includeIgnoreFile()` function reads a file with gitignore-style patterns (such as a `.gitignore`) and returns a config object with the patterns converted to a global ignores object. Pass the absolute path to the ignore file as the first argument:

```js
// eslint.config.js

import { defineConfig, includeIgnoreFile } from "@eslint/config-helpers";
import path from "node:path";

const ignorePath = path.join(import.meta.dirname, ".gitignore");

export default defineConfig([
	includeIgnoreFile(ignorePath, {
		gitignoreResolution: true,
	}),
	// ...
]);
```

#### Options

The second argument is an optional options object:

> - **`gitignoreResolution`** (`boolean`) : Controls how ignore patterns are interpreted.

Reviewer: ```suggestion
- **`gitignoreResolution`** (`boolean`): Controls how ignore patterns are interpreted.
```

Non-blocking, but there's a small typo here.

]);
```

### `includeIgnoreFile()`

> The `includeIgnoreFile()` function reads an ignore file (such as a `.gitignore`) and returns a config object with the patterns converted to a global ignores object. Pass the absolute path to the ignore file as the first argument:

Reviewer 2: ```suggestion
The `includeIgnoreFile()` function reads a file with gitignore-style patterns (such as a `.gitignore`) and returns a config object with the patterns converted to a global ignores object. Pass the absolute path to the ignore file as the first argument:
```

```js
// eslint.config.js

> import { includeIgnoreFile } from "@eslint/config-helpers";

Reviewer 2: ```suggestion
import { defineConfig, includeIgnoreFile } from "@eslint/config-helpers";
```

import path from "node:path";

const ignorePath = path.join(import.meta.dirname, ".gitignore");

export default defineConfig(
	includeIgnoreFile(ignorePath, {
		mode: "gitignore",
	}),
	// ...
);
```

#### Options

The second argument is an optional options object:

- **`mode`** : Controls how ignore patterns are interpreted.

>     - `"eslintignore"` (default) — patterns are resolved relative to the current working directory, matching the behavior of `.eslintignore` files.

Reviewer 2: ```suggestion
    - `"eslintignore"` (default) — patterns are resolved relative to the location of the configuration file.
```

I believe this would be a more correct explanation. Also, I think there's no need to mention legacy .eslintignore files in config-helpers.

Author: Ah - yes, you're right about the CWD vs config file thing, changing that 👍 

On mentioning "eslintignore": Shall we also change the name of the mode then? The reason that I had worded it the way it was was partially to make the default value of `mode: 'eslintignore'` seem like a plausible choice. As a user encountering this as a new API I think I would otherwise be perplexed/annoyed that `mode: "gitignore"` isn't the default. Do you have thoughts as to how to present the default behavior in a way that implies a motivation for it?

Author: Especially on the companion PR (https://github.com/eslint/eslint/pull/20735), I think this will definitely require some thought... currently I have this: https://github.com/eslint/eslint/blob/a25e95f4bec496c129a01a41eb95da024d113cd9/docs/src/use/configure/ignore.md#L220-L307

Reviewer 2: As for the reasoning for the default mode, since users would typically want the other mode, I think it would be best to say that the default mode is the mode it is because it matches the behavior of the first version of this helper, i.e., no particularly good reason other than backwards compatibility. Also, eslint config file and .gitinore file are usually in the same place (project root) so in that case there is no difference between modes.

As for the modes/option name, maybe a boolean "patterns match relative to the location of the ignore file" would be clearer, just not sure how to put this into an option name that wouldn't be overly verbose.

Reviewer 2: Maybe `matchRelativeToIgnoreFile: true`?

By default (`false`), patterns match relative to the location of the configuration file in which they are included, or relative to CWD when the configuration file is used with the `--config` option.

When `true`, patterns always match relative to the location of the ignore file.

Reviewer 3: If we want to use a boolean option, I think it would be better to refer to the usage of the file as a `.gitignore` (maybe `gitignoreResolution`?). This would make it more obvious for user to understand what value to use, instead of focusing on the different base location of relative paths in `.eslintignore` files, which is a poorly documented caveat of the specification.

Reviewer 2: `gitignoreResolution` sounds good to me.

Author: Ok, updated to `gitignoreResolution: boolean`

    - `"gitignore"` — patterns are resolved relative to the ignore file itself (via `basePath`), matching the behavior of `.gitignore` files.
- **`name`** (`string`): A custom name for the resulting config object.

For backwards compatibility with `includeIgnoreFile()` from `@eslint/compat`, passing a string instead of an object as the second argument is treated as equivalent to providing a value for `name`.

#### Multiple files

You can also pass an array of absolute paths to include multiple ignore files at once. In this case an array of config objects is returned:

```js
// eslint.config.js

import { defineConfig, includeIgnoreFile } from "@eslint/config-helpers";
import path from "node:path";

export default defineConfig(
	includeIgnoreFile(
		[
			path.join(import.meta.dirname, ".gitignore"),
			path.join(import.meta.dirname, "packages/lib/.gitignore"),
		],
		{ mode: "gitignore" },
	),
	// ...
);
```

> ### `convertIgnorePatternToMinimatch()`

Reviewer 2: We can remove this from the docs here.

> This is used under the hood by `includeIgnoreFile()` to convert patterns found in eslintignore/gitignore files. You can use it to construct your own ignore objects if you wish:

Reviewer 2: `convertIgnorePatternToMinimatch` was purposely undocumented (see https://github.com/eslint/rewrite/pull/47) as it was intended for internal use only by the `@eslint/migrate-config` package, so I'm not sure if we want to document it now.

Author: Ah - I wasn't aware of that context. I'm happy to do whatever the team prefers here. I guess the levers are:

1. Reexport `convertIgnorePatternToMinimatch()` from `eslint/config`?
2. Export `convertIgnorePatternToMinimatch()` from `@eslint/config-helpers` main entry point?
3. Export `convertIgnorePatternToMinimatch()` from `@eslint/config-helpers` at a use-at-your-own-risk entry opint?
4. (If exported in any way) Document `convertIgnorePatternToMinimatch()`?

I guess we could always start with: don't document and don't export the new copy of `convertIgnorePatternToMinimatch()` in any way. If a use case arises an appropriate export can be added.

Reviewer 3: > `convertIgnorePatternToMinimatch` was purposely undocumented (see #47) as it was intended for internal use only by the `@eslint/migrate-config` package, so I'm not sure if we want to document it now.

In that case, maybe it's not necessary to export `convertIgnorePatternToMinimatch` from `config-helpers`?

Reviewer 2: `@eslint/migrate-config` [uses](https://github.com/eslint/rewrite/blob/cb7e47e77389724d02510d61a636d8a96d1cb31c/packages/migrate-config/src/migrate-config.js#L16) `convertIgnorePatternToMinimatch`, currently from `@eslint/compat`. If we're going to deprecate it in `@eslint/compat`, parhaps it would make sense to switch `@eslint/migrate-config` to using it from `@eslint/config-helpers`?

Reviewer 3: Okay, then I'd say let's switch `@eslint/migrate-config` to depending on `@eslint/config-helpers`, export `convertIgnorePatternToMinimatch` from `@eslint/config-helpers`, and deprecate `convertIgnorePatternToMinimatch` in `@eslint/compat`.

Author: Ok I updated the eslint PR (https://github.com/eslint/eslint/pull/20735) to not reexport `convertIgnorePatternToMinimatch` at `eslint/config` and to not document it.

I updated this PR to remove the documentation for `convertIgnorePatternToMinimatch` as well.

I tried to make the change to migrate-config, but I'm not sure whether it can be done in this PR since `@eslint/migrate-config` will need to reference the next released version of `@eslint/config-helpers`. In any case, the change will be as trivial as 

```diff
diff --git a/packages/migrate-config/package.json b/packages/migrate-config/package.json
index 4a308cf06..80d69ea8f 100644
--- a/packages/migrate-config/package.json
+++ b/packages/migrate-config/package.json
@@ -47,7 +47,7 @@
     "node": "^20.19.0 || ^22.13.0 || >=24"
   },
   "dependencies": {
-    "@eslint/compat": "^2.0.5",
+    "@eslint/config-helpers": "^0.6.0",
     "@eslint/eslintrc": "^3.3.5",
     "camelcase": "^8.0.0",
     "espree": "^10.4.0",
diff --git a/packages/migrate-config/src/migrate-config.js b/packages/migrate-config/src/migrate-config.js
index b1c33fd7b..acc3d55ce 100644
--- a/packages/migrate-config/src/migrate-config.js
+++ b/packages/migrate-config/src/migrate-config.js
@@ -13,7 +13,7 @@ import { Legacy } from "@eslint/eslintrc";
 import camelCase from "camelcase";
 import pluginsNeedingCompat from "./compat-plugins.js";
 import configsNeedingCompat from "./compat-configs.js";
-import { convertIgnorePatternToMinimatch } from "@eslint/compat";
+import { convertIgnorePatternToMinimatch } from "@eslint/config-helpers";
 import * as espree from "espree";
```

(assuming 0.6.0 is the next release for config-helpers)

Also - I'm thinking maybe `migrate-config` should be updated to use the new version of `includeIgnoreFile` in its codegen (rather than the now-deprecated version in `@eslint/compat`) but this should probably also wait until the config-helpers package is actuallly released.

Reviewer 2: > Also - I'm thinking maybe `migrate-config` should be updated to use the new version of `includeIgnoreFile` in its codegen (rather than the now-deprecated version in `@eslint/compat`) but this should probably also wait until the config-helpers package is actuallly released.

We could consider updating it to use `includeIgnoreFile` from `eslint/config` after releasing `eslint` with changes from https://github.com/eslint/eslint/pull/20735.
